src/SparkConduit.sol

```diff
@@ -0,0 +1,317 @@
+    {
+        uint256 liquidityAvailable
+            = IERC20(asset).balanceOf(pool.getReserveData(asset).aTokenAddress);
+
+        amount = liquidityAvailable < maxAmount ? liquidityAvailable : maxAmount;
+
+        // Normally you should update local state first for re-entrancy,
+        // but we need an update-to-date liquidity index for that
+        amount = pool.withdraw(asset, amount, address(this));
+
+        // Convert asset amount to shares
+        uint256 liquidityIndex = pool.getReserveData(asset).liquidityIndex;
+        uint256 shares         = amount.rayDiv(liquidityIndex);
+
+        uint256 deposits    = assets[asset].positions[ilk].deposits;
+        uint256 withdrawals = assets[asset].positions[ilk].withdrawals;
+
+        if (deposits < shares) {
+            uint256 toReturnShares;
+
+            unchecked {
+                toReturnShares = shares - deposits;
+            }
+
+            uint256 toReturn = toReturnShares.rayMul(liquidityIndex);
+            amount -= toReturn;
+            shares = deposits;
+
+            // Return the excess to the pool as it's other user's deposits
+            // Round against the user that is withdrawing
+            pool.supply(asset, toReturn, address(this), 0);
+        }
+
+        assets[asset].positions[ilk].deposits -= shares;
+        assets[asset].totalDeposits           -= shares;
+
+        if (withdrawals > 0) {
+            if (shares <= withdrawals) {
+                assets[asset].positions[ilk].withdrawals -= shares;
+                assets[asset].totalWithdrawals           -= shares;
+            } else {
+                assets[asset].positions[ilk].withdrawals = 0;
+                assets[asset].totalWithdrawals           -= withdrawals;
+            }
+        }
+
+        address destination = RegistryLike(registry).buffers(ilk);
+
+        IERC20(asset).transfer(destination, amount);
+
+        emit Withdraw(ilk, asset, destination, amount);
+    }
+
+    /// @inheritdoc IAllocatorConduit
+    function maxDeposit(bytes32, address asset) public view returns (uint256 maxDeposit_) {
+        if (assets[asset].enabled) {
+            // Note: Purposefully ignoring any potental supply cap limits on Spark
+            return type(uint256).max;
+        } else {
+            return 0;
+        }
+    }
+
+    /// @inheritdoc IAllocatorConduit
+    function maxWithdraw(bytes32 ilk, address asset) public view returns (uint256 maxWithdraw_) {
+        DataTypes.ReserveData memory reserveData = pool.getReserveData(asset);
+        maxWithdraw_ = assets[asset].positions[ilk].deposits.rayMul(reserveData.liquidityIndex);
+
+        uint256 liquidityAvailable = IERC20(asset).balanceOf(reserveData.aTokenAddress);
+        if (maxWithdraw_ > liquidityAvailable) maxWithdraw_ = liquidityAvailable;
+    }
+
+    /// @inheritdoc ISparkConduit
+    function requestFunds(bytes32 ilk, address asset, uint256 amount) external ilkAuth(ilk) {
```

Comment:
I see that this request is only possible if liquidityAvailable is zero, a few questions:

1. Couldn't this be frontrun by adding 1 wei of `asset` to the aToken as a DoS vector?
2. I assume this is to increase rates in Spark, but I don' see any logic related to `totalWithdrawals` in `calculateInterestRates` or `recompute`, maybe I'm missing something?

Reply:
Ah I see it's through `targetDebt`, nvm

Reply:
Good point about 1, but any call to requestFunds likely needs to be done atomically anyways. Some external contract should be combining the withdraw and requestFunds. Maybe it makes sense to have a function inside SparkConduit for convenience?

Reply:
Made a note to address in next PR